Update OpenSSH docs with agentless information (#23359)
gravitational/teleport: docs/pages/server-access/guides/openssh.mdx

Using Teleport and OpenSSH has the advantage of getting you up
and running, but in the long run, we would recommend replacing `sshd` with `teleport`.
`teleport` SSH servers have support for multiple features that are incompatible with OpenSSH:

> - [Dynamically updating node labels from the output of configured commands](../../management/admin/labels.mdx)

Reviewer: Labels are Teleport specific, so I'm not sure why this would be meaningful to an OpenSSH user. Is the expectation here that OpenSSH users would understand what Teleport labels are?

Author: My thoughts were having an explicit list of what is and isn't supported with agentless nodes would be useful to set user's expectations, what do you think, is that not necessary?

- [Session recording without SSH connection termination](recording-proxy-mode.mdx)

> - [Advanced session recording](bpf-session-recording.mdx) and [restricting outbound network connections in SSH sessions](restricted-session.mdx)

Reviewer: Why are these two links included in the same bullet?

Author: No good reason I guess, they both use BPF but users don't care about that. Split into two bullets

Using Teleport and OpenSSH has the advantage of getting you up
and running, but in the long run, we would recommend replacing `sshd` with `teleport`.
`teleport` SSH servers have support for multiple features that are incompatible with OpenSSH:

> - [Creating ephemeral Unix users when logging into a node](host-user-creation.mdx)

Reviewer 2: Technically I think you can do it with plain OpenSSH and PAM so it's not like this is Teleport-exclusive feature.

> - [Dynamically updating node labels from the output of configured commands](../../management/admin/labels.mdx)

Reviewer: ```suggestion
- RBAC and resource filtering based on [dynamically updated labels](../../management/admin/labels.mdx)
```
Suggested so there's a clear benefit to this

Author: Thanks, done

- [Dynamically updating node labels from the output of configured commands](../../management/labels.mdx)
- [Session recording without SSH connection termination](recording-proxy-mode.mdx)

> - [Logging of *all* commands](bpf-session-recording.mdx) and [restricting outbound network connections in SSH sessions](restricted-session.mdx)

Reviewer 2: I'd just call it what we call it everywhere else, "advanced session recording".
```suggestion
- [Advanced session recording](bpf-session-recording.mdx) and [restricting outbound network connections in SSH sessions](restricted-session.mdx)
```

Author: Done

Using Teleport and OpenSSH has the advantage of getting you up
and running, but in the long run, we would recommend replacing `sshd` with `teleport`.

> `teleport` SSH servers have support for multiple features that are incompatible with OpenSSH:

Reviewer: Looking at this list, I think we should choose three features and add a sentence to describe each one. Otherwise, it's not really possible to understand what these terms describe without clicking on the links.

Author: I kept all of them, I think it's important for users to know what agentless nodes can and can't do compared to Teleport nodes. If you still think the list should be shortened I can remove some though.

- [Automated host user creation](host-user-creation.mdx)

> - [Dynamic labels](../../management/labels.mdx)

Reviewer: Is the ability to add labels a feature in itself, or is it more about what you can do with the labels? Since this list is already pretty long, I think we can remove this item.

Author: Just dynamic labels, regular labels can be added to agentless nodes. I'll update to reflect that.

- [Session recording without SSH connection termination](recording-proxy-mode.mdx)

> - [Enhanced session recording](bpf-session-recording.mdx)

Reviewer: Would an OpenSSH user know what this is?

Author: Updated with a better description

> - [Restricted networking](restricted-session.mdx)

Reviewer: I don't think this is self explanatory.

Author: Updated with better description

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- A Linux host with the OpenSSH server `sshd` version 7.4 or above installed,
  but not Teleport. The SSH port on this host must be open to traffic from the
  Teleport Proxy Service host.
- (!docs/pages/includes/tctl.mdx!)

## Step 1/5. Add a node resource to your Teleport cluster

When you request an SSH connection to a OpenSSH node, Teleport needs to be able
to find the node's IP address so it can establish a connection to it.

Declare a `node` resource so Teleport knows how to reach your OpenSSH server.
On your local machine, create a file called `openssh-node-resource.yaml` with the following content:

```yaml
kind: node
version: v2
sub_kind: openssh
metadata:
  name: a100fdd0-52db-4eca-a7ab-c3afa7a1564a
  labels:
    env: prod
spec:
  addr: 1.2.3.4:22
  hostname: openssh-node
```

> The fields `spec.addr` and `spec.hostname` are mandatory, but `metadata.labels` is not.
> The `metadata.name` field isn't mandatory, but setting it here will save you some work later.

Reviewer: ```suggestion
`spec.addr` and `spec.hostname` are mandatory. Assign `spec.addr` to the address and port of your node and `spec.hostname` to the name of the node as you would like users to see it in Teleport.

 The `metadata.labels` field labels the SSH Service instance so you can apply RBAC rules to it. 

The `metadata.name` field isn't mandatory, but setting it here will save you some work later.
```

I'm not sure we should assume that a reader will correctly guess the meanings of configuration fields.

Author: Yeah that's much better, done

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- A Linux host with the OpenSSH server `sshd` version 7.4 or above installed,
  but not Teleport. The SSH port on this host must be open to traffic from the
  Teleport Proxy Service host.

(!docs/pages/includes/tctl.mdx!)

## Step 1/5. Add a node resource to your Teleport cluster

When you request an SSH connection to a OpenSSH node, Teleport needs to be able
to find the node's IP address so it can establish a connection to it.

Declare a `node` resource so Teleport knows how to reach your OpenSSH server.
On your local machine, create a file called `openssh-node-resource.yaml` with the following content:

```yaml
kind: node
sub_kind: openssh
metadata:
  labels:
    env: prod
spec:
  addr: 1.2.3.4:22

>   hostname: openssh-node

Reviewer 2: Is `hostname` mandatory? Let's mention that.

Author: Done

> version: v2

Reviewer 2: Nit: Can you move `version` right after `kind`? It looks weird at the bottom of the resource.

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- A Linux host with the OpenSSH server `sshd` version 7.4 or above installed,
  but not Teleport. The SSH port on this host must be open to traffic from the
  Teleport Proxy Service host.

(!docs/pages/includes/tctl.mdx!)

## Step 1/5. Add a node resource to your Teleport cluster

> 

Reviewer: > Teleport needs to know how to reach your OpenSSH server

I think it would help to clarify the sentence to add a short paragraph at the top of this section indicating what is happening architecturally in the setup documented here. What is the Teleport cluster doing with the `node` resource?

Author: Good idea, done

> Teleport needs to know how to reach your OpenSSH server, so a node resource must
> be created. On your local machine, create a file `openssh-node-resource.yaml`:

Reviewer: ```suggestion
Declare a `node` resource so Teleport knows how to reach your OpenSSH server. On your local machine, create a file called `openssh-node-resource.yaml` with the following content:
```

```yaml
kind: node
sub_kind: openssh
metadata:
  labels:
    env: prod
spec:

>   addr: 1.2.3.4:22

Reviewer: Will this `addr` apply to the environments of all users following this guide? Ideally, it should be possible for a user to go through the steps themselves and achieve a working result.

Author: This document already used example values for a few things so that's what I did too. Would you prefer I use placeholders instead such as `<ADDR>`?

stdout:

```code

> $ tctl auth export --type=openssh | sed "s/cert-authority\ //"

Reviewer 3: FYI, @marcoandredinis made it possible to curl this info without requiring you to be logged in. It's a bit more convenient because you can run curl directly from the openssh node rather than transferring a file to the node.

Author: Looks like @marcoandredinis also opened a PR to update the same doc page: https://github.com/gravitational/teleport/pull/23560 kinda a chicken and egg problem...

Reviewer 4: Feel free to keep it as it
I'll update it after you merge your PR :+1:

Author: Sweet thank you!

### Issue a host certificate

> When creating host certificates, it is important to specify all the domain names

Reviewer: - We're no longer capitalizing "Node" in the docs to refer to an SSH Service instance
- I'm not sure if this paragraph belongs in this location. We haven't shown the user how to create a host certificate yet, so they might get the impression that we're leaving them to figure this out for themselves. Can we put these instructions by the `tctl auth sign` instructions?

Author: Yeah that's a better place for it, done

and addresses that refer to your Node. If you try to connect to a Node with a 
name or address that was not specified when creating it's host certificate,
Teleport will reject the SSH connection.

> #### Finding your Node's UUID

Author: Done

> When you created a node resource earlier, a unique UUID was generated for that

Reviewer: ```suggestion
When you created a `node` resource earlier, the Teleport Auth Service generated a universal unique identifier (UUID) for that
```
- Using code style when referring to a configuration resource
- Using active voice for clarity
- Removing the repetition in "unique UUID"

Author: Done

> node. This is so multiple nodes with the same hostname can be uniquely identified.

Reviewer: Can be uniquely identified by what? We should use active voice here so this is clearer ("x uses the UUID to uniquely identify multiple nodes with the same hostname" or something like that).

Author: Done

> This must be added to the host certificate. To find your Node's UUID, first

Reviewer: same as above re: "Node"

> determine if it's hostname is unique:

Reviewer: ```suggestion
determine if its hostname is unique:
```

```code
$ tctl get node/openssh-node --format text
```

If only one node is displayed and you have `jq` installed, you can run the

> following command to get your Node's UUID:

Reviewer: same as above re: "Node"

```code
$ tctl get node/openssh-node --format=json | jq -r ".[0].metadata.name"
```

> Otherwise, find your Node's UUID in the `metadata.name` field of the yaml

Reviewer: same as above re: "Node"

> Otherwise, find your Node's UUID in the `metadata.name` field of the yaml

Reviewer: ```suggestion
Otherwise, find your Node's UUID in the `metadata.name` field of the YAML
```

universal unique identifier (UUID) for that node. Teleport Proxy Services uses
the UUID to differentiate nodes with the same hostname, so it be added to the host certificate.
To find your node's UUID, first determine if its hostname is unique:

```code
$ tctl get node/openssh-node --format text
```

If only one node is displayed and you have `jq` installed, you can run the
following command to get your node's UUID:

```code
$ tctl get node/openssh-node --format=json | jq -r ".[0].metadata.name"
```

Otherwise, find your node's UUID in the `metadata.name` field of the YAML
output of this command:

```code
$ tctl get node/openssh-node
```

#### Create the host certificate

When creating host certificates, it is important to specify all the domain names
and addresses that refer to your node. If you try to connect to a node with a 
name or address that was not specified when creating it's host certificate,
Teleport will reject the SSH connection.

On your local machine, assign the IP address, fully qualified domain name of
your node, and the node's UUID to an environment variable. If you won't be
connecting to your node with its hostname, you can safely omit it.

```code

> $ ADDR=1.2.3.4,openssh-node,a100fdd0-52db-4eca-a7ab-c3afa7a1564a

Reviewer 2: Is finding/including the node's UUID required?

Reviewer 2: I guess my question was more of "why" is it required? Is it required currently for OpenSSH nodes too? Just thinking that having to find out this ID might complicate automation for users.

Author: Unregistered OpenSSH nodes don't have UUIDs assigned to them, as they are not a registered Teleport resource, so that was not required.

Reviewer 2: @capnspacehook When you're creating the OpenSSH server resource, you can assign the UUID yourself, correct? Think it would make sense to mention this in the docs? Then users wouldn't need to fetch the resource back after creating it at least.

Author: True, fixed

## Revoke an SSH certificate

To revoke the current Teleport CA and generate a new one, run `tctl auth rotate`. Unless you've highly automated your

> infrastructure, we would suggest you proceed with caution as this will invalidate the openssh

Reviewer: ```suggestion
infrastructure, we would suggest you proceed with caution as this will invalidate the OpenSSH
```
